docs/src/pages/components/data-grid/pagination/CursorPaginationGrid.tsx

```diff
@@ -0,0 +1,113 @@
+import * as React from 'react';
+import {
+  GridRowsProp,
+  DataGrid,
+  GridPageChangeParams,
+} from '@material-ui/data-grid';
+import {
+  useDemoData,
+  GridData,
+  DemoDataOptions,
+  DataRowModel,
+} from '@material-ui/x-grid-data-generator';
+
+interface CursorBasedGridData extends GridData {
+  cursors: string[];
+}
+
+interface CursorBasedGridResponse {
+  rows: DataRowModel[];
+  nextCursor: string;
+}
+
+function loadServerRows(
+  cursor: string | null | undefined,
+  data: CursorBasedGridData,
+): Promise<CursorBasedGridResponse> {
+  return new Promise<CursorBasedGridResponse>((resolve) => {
+    setTimeout(() => {
+      const start = cursor ? data.cursors.indexOf(cursor) + 1 : 0;
+      const end = start + 5;
+      const rows = data.rows.slice(start, end);
+
+      resolve({ rows, nextCursor: data.cursors[end] });
+    }, Math.random() * 500 + 100); // simulate network latency
+  });
+}
+
+const useCursorBasedDemoData = (options: DemoDataOptions) => {
+  const response = useDemoData(options);
+
+  const data = React.useMemo(
+    () => ({
+      ...response.data,
+      cursors: response.data.rows.map(() => Math.random().toString()),
+    }),
+    [response.data],
+  );
+
+  return {
+    ...response,
+    data,
+  };
+};
+
+export default function ServerPaginationGrid() {
+  const { data } = useCursorBasedDemoData({
+    dataSet: 'Commodity',
+    rowLength: 100,
+    maxColumns: 6,
+  });
+
+  const pagesNextCursor = React.useRef<{ [page: number]: string }>({});
+
+  const [rows, setRows] = React.useState<GridRowsProp>([]);
+  const [page, setPage] = React.useState(0);
+  const [loading, setLoading] = React.useState<boolean>(false);
+
+  const handlePageChange = (params: GridPageChangeParams) => {
+    setPage(params.page);
+  };
```

Comment:
We might not want to allow going to a page where we don't have loaded yet the next page cursor. Meaning, if you go quickly from page 0 to page 2. We have no ways to load the data, as we need page 1 to now the next cursor.
```suggestion
  const handlePageChange = (params: GridPageChangeParams) => {
    // We have the cursor, we can allow the page transition.
    if (pagesNextCursor.current[params.page - 1]) {
      setPage(params.page);
    }
  };
```

Reply:
@tobiasbueschel Would this new demo be enough for your use case? (we won't change the API just yet). How do you solve this problem with react-query?